File: algorithms/python/011_maxArea.py
# -*- encoding: utf-8 -*-

class Solution:
    def maxArea(self, height):
        """
        :param height:
        :return:
        """
        # 思路一.双循环,遍历所有可能的两两组合,时间复杂度太高O(n2)
        # 思路二.需要证明
        MAX = 0
        x = len(height) - 1
        y = 0
        while x != y:
            if height[x] > height[y]:
                area = height[y] * (x - y)
                y += 1
            else:
                area = height[x] * (x - y)
                x -= 1
            MAX = max(MAX, area)

        return MAX

File: algorithms/cpp/011_maxArea.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int maxArea(vector<int>& height)
{
    if(height.size() <= 1) return 0;
    int start = 0, end = height.size() - 1;
    int max_area = 0;
    while(start < end)
    {
        int h = min(height[start], height[end]);
        max_area = max(max_area, (end - start) * h);
        while(height[start] <= h) start++;
        while(height[end] <= h) end--;
    }
    return max_area;
}

int main()
{
    int n;
    cin >> n;
    vector<int> height;
    for(int i = 0; i < n; ++i)
    {
        int tmp;
        cin >> tmp;
        height.push_back(tmp);
    }
    int res = maxArea(height);
    return res;
}

